File: core/src/main/java/com/oneandone/typedrest/GenericCollectionEndpoint.java
package com.oneandone.typedrest;

import java.io.*;
import java.util.Collection;
import java.util.List;
import java.util.Optional;
import org.apache.http.*;

/**
 * REST endpoint that represents a collection of <code>TEntity</code>s as
 * <code>TElementEndpoint</code>s.
 *
 * Use the more constrained {@link CollectionEndpoint} when possible.
 *
 * @param <TEntity> The type of entity the endpoint represents.
 * @param <TElementEndpoint> The specific type of {@link ElementEndpoint} to
 * provide for individual <code>TEntity</code>s.
 */
public interface GenericCollectionEndpoint<TEntity, TElementEndpoint extends Endpoint>
        extends Endpoint {

    /**
     * Returns the type of entity the endpoint represents.
     *
     * @return The class type.
     */
    Class<TEntity> getEntityType();

    /**
     * Returns a {@link ElementEndpoint} for a specific child element of this
     * collection.
     *
     * @param id The ID identifying the entity in the collection.
     * @return An {@link ElementEndpoint} for a specific element of this
     * collection.
     */
    TElementEndpoint get(String id);

    /**
     * Returns a {@link ElementEndpoint} for a specific child element of this
     * collection.
     *
     * @param entity An existing entity to extract the ID from.
     * @return An {@link ElementEndpoint} for a specific element of this
     * collection.
     */
    TElementEndpoint get(TEntity entity);

    /**
     * Shows whether the server has indicated that
     * {@link #readAll() is currently allowed.
     *
     * Uses cached data from last response.
     *
     * @return An indicator whether the method is allowed. If no request has
     * been sent yet or the server did not specify allowed methods
     * {@link Optional#empty()} is returned.
     */
    Optional<Boolean> isReadAllAllowed();

    /**
     * Returns all <code>TEntity</code>s.
     *
     * @return All <code>TEntity</code>s.
     * @throws IOException Network communication failed.
     * @throws IllegalArgumentException {@link HttpStatus#SC_BAD_REQUEST}
     * @throws IllegalAccessException {@link HttpStatus#SC_UNAUTHORIZED} or
     * {@link HttpStatus#SC_FORBIDDEN}
     * @throws FileNotFoundException {@link HttpStatus#SC_NOT_FOUND} or
     * {@link HttpStatus#SC_GONE}
     * @throws IllegalStateException {@link HttpStatus#SC_CONFLICT}
     * @throws RuntimeException Other non-success status code.
     */
    List<TEntity> readAll()
            throws IOException, IllegalArgumentException, IllegalAccessException, FileNotFoundException, IllegalStateException;

    /**
     * Shows whether the server has indicated that {@link #readRange()} is
     * allowed.
     *
     * Uses cached data from last response.
     *
     * @return An indicator whether the method is allowed. If no request has
     * been sent yet {@link Optional#empty()} is returned.
     */
    Optional<Boolean> isReadRangeAllowed();

    /**
     * Returns all <code>TElementEndpoint</code>s within a specific range of the
     * set.
     *
     * @param from The index of the first element to return. <code>null</code>
     * to use <code>to</code> to specify a start point counting from the end of
     * the set.
     * @param to The index of the last element to return. Alternatively the
     * index of the first element to return counting from the end of the set if
     * <code>from</code> is <code>null</code>. <code>null</code> to read to the
     * end.
     * @return A subset of the <code>TElementEndpoint</code>s and the range they
     * come from. May not exactly match the request range.
     * @throws IOException Network communication failed.
     * @throws IllegalArgumentException {@link HttpStatus#SC_BAD_REQUEST}
     * @throws IllegalAccessException {@link HttpStatus#SC_UNAUTHORIZED} or
     * {@link HttpStatus#SC_FORBIDDEN}
     * @throws FileNotFoundException {@link HttpStatus#SC_NOT_FOUND} or
     * {@link HttpStatus#SC_GONE}
     * @throws IllegalStateException The requested range is not satisfiable.
     * @throws RuntimeException Other non-success status code.
     */
    PartialResponse<TEntity> readRange(Long from, Long to)
            throws IOException, IllegalArgumentException, IllegalAccessException, FileNotFoundException, IllegalStateException;

    /**
     * Shows whether the server has indicated that
     * {@link #create(java.lang.Object)} is currently allowed.
     *
     * Uses cached data from last response.
     *
     * @return An indicator whether the method is allowed. If no request has
     * been sent yet or the server did not specify allowed methods
     * {@link Optional#empty()} is returned.
     */
    Optional<Boolean> isCreateAllowed();

    /**
     * Creates a new <code>TEntity</code>.
     *
     * @param entity The new <code>TEntity</code>.
     * @return The newly created <code>TEntity</code>; may be <code>null</code>
     * if the server deferred creating the resource.
     *
     * @throws IOException Network communication failed.
     * @throws IllegalArgumentException {@link HttpStatus#SC_BAD_REQUEST}
     * @throws IllegalAccessException {@link HttpStatus#SC_UNAUTHORIZED} or
     * {@link HttpStatus#SC_FORBIDDEN}
     * @throws FileNotFoundException {@link HttpStatus#SC_NOT_FOUND} or
     * {@link HttpStatus#SC_GONE}
     * @throws IllegalStateException {@link HttpStatus#SC_CONFLICT}
     * @throws RuntimeException Other non-success status code.
     */
    TElementEndpoint create(TEntity entity)
            throws IOException, IllegalArgumentException, IllegalAccessException, FileNotFoundException, IllegalStateException;

    /**
     * Shows whether the server has indicated that
     * {@link #createAll(java.lang.Iterable)} is currently allowed.
     *
     * Uses cached data from last response.
     *
     * @return An indicator whether the method is allowed. If no request has
     * been sent yet or the server did not specify allowed methods
     * {@link Optional#empty()} is returned.
     */
    Optional<Boolean> isCreateAllAllowed();

    /**
     * Creates multiple new <code>TEntity</code>s.
     *
     * @param entities The new <code>TEntity</code>s.
     *
     * @throws IOException Network communication failed.
     * @throws IllegalArgumentException {@link HttpStatus#SC_BAD_REQUEST}
     * @throws IllegalAccessException {@link HttpStatus#SC_UNAUTHORIZED} or
     * {@link HttpStatus#SC_FORBIDDEN}
     * @throws FileNotFoundException {@link HttpStatus#SC_NOT_FOUND} or
     * {@link HttpStatus#SC_GONE}
     * @throws IllegalStateException {@link HttpStatus#SC_CONFLICT}
     * @throws RuntimeException Other non-success status code.
     */
    void createAll(Iterable<TEntity> entities)
            throws IOException, IllegalArgumentException, IllegalAccessException, FileNotFoundException, IllegalStateException;

    /**
     * Shows whether the server has indicated that
     * {@link #setAll(java.util.Collection)} is currently allowed.
     *
     * Uses cached data from last response.
     *
     * @return An indicator whether the method is allowed. If no request has
     * been sent yet or the server did not specify allowed methods
     * {@link Optional#empty()} is returned.
     */
    Optional<Boolean> isSetAllAllowed();

    /**
     * Replaces the entire content of the collection with new
     * <code>TEntity</code>s.
     *
     * @param entities >The new set of <code>TEntity</code>s the collection
     * shall contain.
     * @throws IOException Network communication failed.
     * @throws IllegalArgumentException {@link HttpStatus#SC_BAD_REQUEST}
     * @throws IllegalAccessException {@link HttpStatus#SC_UNAUTHORIZED} or
     * {@link HttpStatus#SC_FORBIDDEN}
     * @throws FileNotFoundException {@link HttpStatus#SC_NOT_FOUND} or
     * {@link HttpStatus#SC_GONE}
     * @throws IllegalStateException The entities have changed since they were
     * last retrieved with {@link #readAll()}. Your changes were rejected to
     * prevent a lost update.
     * @throws RuntimeException Other non-success status code.
     */
    void setAll(Collection<TEntity> entities)
            throws IOException, IllegalArgumentException, IllegalAccessException, FileNotFoundException;
}

File: typedrest/src/main/kotlin/net/typedrest/endpoints/generic/GenericCollectionEndpoint.kt
package net.typedrest.endpoints.generic

import net.typedrest.errors.ConflictException
import net.typedrest.http.*
import net.typedrest.errors.*

/**
 * Endpoint for a collection of [TEntity]s addressable as [TElementEndpoint]s.
 *
 * Use the more constrained [CollectionEndpoint] when possible.
 *
 * @param TEntity The type of individual elements in the collection.
 * @param TElementEndpoint The type of [ElementEndpoint] to provide for individual [TEntity]s.
 */
interface GenericCollectionEndpoint<TEntity, out TElementEndpoint : ElementEndpoint<TEntity>> {
    /**
     * Returns an [ElementEndpoint] for a specific child element.
     *
     * @param entity An existing entity to extract the ID from.
     * @return The [TElementEndpoint] for the specified entity.
     */
    operator fun get(entity: TEntity): TElementEndpoint

    /**
     * Shows whether the server has indicated that [readAll] is currently allowed.
     *
     * Uses cached data from last response.
     *
     * @return true if the method is allowed, false if the method is not allowed, null If no request has been sent yet or the server did not specify allowed methods.
     */
    val readAllAllowed: Boolean?

    /**
     * Returns all entities in the collection.
     *
     * @throws AuthenticationException when the server responds with [HttpStatusCode.Unauthorized].
     * @throws AuthorizationException when the server responds with [HttpStatusCode.Forbidden].
     * @throws NotFoundException when the server responds with [HttpStatusCode.NotFound] or [HttpStatusCode.Gone].
     * @throws HttpException for other non-success status codes.
     * @return The list of all [TEntity].
     */
    fun readAll(): List<TEntity>

    /**
     * Shows whether the server has indicated that [readRange] is allowed.
     *
     * Uses cached data from last response.
     *
     * @return An indicator whether the method is allowed. If no request has been sent yet.
     */
    val readRangeAllowed: Boolean?

    /**
     * Returns all entities within a specific range of the collection.
     *
     * @param from The position at which to start sending data.
     * @param to The position at which to stop sending data.
     * @return A [PartialResponse] containing a subset of the entities and the range they come from.
     * @throws AuthenticationException when the server responds with [HttpStatusCode.Unauthorized].
     * @throws AuthorizationException when the server responds with [HttpStatusCode.Forbidden].
     * @throws NotFoundException when the server responds with [HttpStatusCode.NotFound] or [HttpStatusCode.Gone].
     * @throws ConflictException if the requested range is not satisfiable.
     * @throws HttpException for other non-success status codes.
     */
    fun readRange(from: Long?, to: Long?): PartialResponse<TEntity>

    /**
     * Shows whether the server has indicated that [create] is currently allowed.
     *
     * Uses cached data from last response.
     *
     * @return true if the method is allowed, false if the method is not allowed, null If no request has been sent yet or the server did not specify allowed methods.
     */
    val createAllowed: Boolean?

    /**
     * Adds an entity as a new element to the collection.
     *
     * @param entity The new entity.
     * @return An endpoint for the newly created entity; `null` if the server returned neither a "Location" header nor an entity with an ID in the response body.
     * @throws BadRequestException when the server responds with [HttpStatusCode.BadRequest].
     * @throws AuthenticationException when the server responds with [HttpStatusCode.Unauthorized].
     * @throws AuthorizationException when the server responds with [HttpStatusCode.Forbidden].
     * @throws ConflictException for when the server responds with [HttpStatusCode.Conflict].
     * @throws HttpException for other non-success status codes.
     */
    fun create(entity: TEntity): TElementEndpoint?

    /**
     * Shows whether the server has indicated that [createAll] is currently allowed.
     *
     * Uses cached data from last response.
     *
     * @return An indicator whether the verb is allowed. If no request has been sent yet or the server did not specify allowed verbs `null` is returned.
     */
    val createAllAllowed: Boolean?

    /**
     * Adds (or updates) multiple entities as elements in the collection.
     *
     * Uses a link with the relation type `bulk` to determine the URI to POST to. Defaults to the relative URI `bulk`.
     *
     * @param entities The entities to create or modify.
     * @throws BadRequestException when the server responds with [HttpStatusCode.BadRequest].
     * @throws AuthenticationException when the server responds with [HttpStatusCode.Unauthorized].
     * @throws AuthorizationException when the server responds with [HttpStatusCode.Forbidden].
     * @throws ConflictException for when the server responds with [HttpStatusCode.Conflict].
     * @throws HttpException for other non-success status codes.
     */
    fun createAll(entities: Iterable<TEntity>)

    /**
     * Shows whether the server has indicated that [setAll] is currently allowed.
     *
     * Uses cached data from last response.
     *
     * @return An indicator whether the verb is allowed. If no request has been sent yet or the server did not specify allowed verbs `null` is returned.
     */
    val setAllAllowed: Boolean?

    /**
     * Replaces the entire content of the collection with new entities.
     *
     * @param entities The new set of entities the collection shall contain.
     * @throws BadRequestException when the server responds with [HttpStatusCode.BadRequest].
     * @throws AuthenticationException when the server responds with [HttpStatusCode.Unauthorized].
     * @throws AuthorizationException when the server responds with [HttpStatusCode.Forbidden].
     * @throws ConflictException when the entity has changed since it was last retrieved with [readAll]. Your changes were rejected to prevent a lost update.
     * @throws HttpException for other non-success status codes.
     */
    fun setAll(entities: Iterable<TEntity>)
}
